pandas/core/indexes/datetimes.py

```diff
@@ -30,6 +32,29 @@
 from pandas.core.indexes.extension import inherit_names
 from pandas.core.tools.times import to_time
 
+if TYPE_CHECKING:
+    from pandas import DataFrame, PeriodIndex, TimedeltaIndex
+
+
+def dispatch_to_array_and_wrap(delegate) -> Callable[[F], F]:
+    def decorator(func: F) -> F:
+        name = func.__name__
```

Comment:
Doesn't functools.wraps already copy over the `__name__` and `__doc__`? I think you might be able to reduce the layers of wrapping by one

Reply:
this to copy doc from delegate not func.

wraps is to copy __name__ (the code here does not set the __name__ attribute) and the signature for the help/docs

can always add the doc decorator to each method individually and no need for the decorator factory. I did this originally but thought it looked messy with two decorators and repetition.